keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -607,28 +617,251 @@ Setup a guaranteed class Pod with two containers (c1 & c2).
 
 TODO: Identify more cases
 
-## Graduation Criteria
+### Graduation Criteria
 
-### Alpha
+#### Alpha
 - In-Place Pod Resouces Update functionality is implemented for running Pods,
 - LimitRanger and ResourceQuota handling are added,
 - Resize Policies functionality is implemented,
 - Unit tests and E2E tests covering basic functionality are added,
 - E2E tests covering multiple containers are added.
 
-### Beta
+#### Beta
 - VPA alpha integration of feature completed and any bugs addressed,
 - E2E tests covering Resize Policy, LimitRanger, and ResourceQuota are added,
 - Negative tests are identified and added.
 - A "/resize" subresource is defined and implemented.
 - Pod-scoped resources are handled if that KEP is past alpha
 
-### Stable
+#### Stable
 - VPA integration of feature moved to beta,
 - User feedback (ideally from atleast two distinct users) is green,
 - No major bugs reported for three months.
 - Pod-scoped resources are handled if that KEP is past alpha
 
+### Upgrade / Downgrade Strategy
+Scheduler and API server should be updated before Kubelets in that order.
+Kubelet and the runtime versions should use the same CRI version in lock-step.
+Upgrade involves draining all pods from a node, installing a CRI runtime with this
+version of the API and update to a matching kubelet and making node schedulable again.
+Downgrade involves doing the above in reverse.
+
+### Version Skew Strategy
+Kubelet and the CRI runtime versions are expected to match so we don't have to worry about.
+Previous versions of clients that are unaware of the new ResizePolicy fields would set them
+to nil. API server mutates such updates by copying non-nil values from old Pod to current Pod
```

Comment:
What happens when the API server supports the feature but the kubelets do not? Keep in mind that kubelets can be up to n-2 versions behind the API server.

Reply:
In that case feature would not work as expected. Enabling the feature-flag will allow apiserver to accept PATCH to PodSpec.Containers[].Resources field, but kubelet will interpret it as a change to the container definition and restart the container. What's worse is that on restart it would apply the new resources spec and thus it may oversubscribe the node. 

I'm not sure if there's a good way to mitigate this without ugly stop-gap apiserver hacks to check node version. Is stretching out alpha to n+2 a reasonable approach?

Reply:
> In that case feature would not work as expected. Enabling the feature-flag will allow apiserver to accept PATCH to PodSpec.Containers[].Resources field, but kubelet will interpret it as a change to the container definition and restart the container. What's worse is that on restart it would apply the new resources spec and thus it may oversubscribe the node.
> 
> I'm not sure if there's a good way to mitigate this without ugly stop-gap apiserver hacks to check node version. Is stretching out alpha to n+2 a reasonable approach?

This is why we ask these questions, to catch them before we've implemented it without a strategy :)

I don't think we should delay beta, but this would delay GA, since we'd need to ensure that the feature gate was on in kubelets before removing it in the API server.

Reply:
Good catch indeed! I believe we turn the feature gate on in beta. So, is a surprise such as this OK in beta?

Reply:
It shouldn't be a surprise, but rather, a documented failure mode.